apollo-router/src/axum_http_server_factory.rs

```diff
@@ -0,0 +1,1031 @@
+                                    std::io::ErrorKind::UnexpectedEof => todo!(),
+                                    std::io::ErrorKind::OutOfMemory => todo!(),*/
+                                    _ => {
+                                        continue;
+                                    }
+
+                                }
+                            }
+                        }
+                    }
+                }
+
+                // the shutdown receiver was triggered so we break out of
+                // the server loop, tell the currently active connections to stop
+                // then return the TCP listen socket
+                connection_shutdown.notify_waiters();
+                listener
+            };
+
+            // Spawn the server into a runtime
+            let server_future = tokio::task::spawn(server)
+                .map_err(|_| FederatedServerError::HttpServerLifecycleError)
+                .boxed();
+
+            Ok(HttpServerHandle::new(
+                shutdown_sender,
+                server_future,
+                actual_listen_address,
+            ))
+        })
+    }
+}
+
+#[derive(Debug)]
+struct CustomRejection {
+    #[allow(dead_code)]
+    msg: String,
+}
+
+async fn custom_plugin_handler(
+    Host(host): Host,
+    request: Request<Body>,
+    handler: Handler,
+) -> impl IntoResponse {
+    let (mut head, body) = request.into_parts();
+    let body = hyper::body::to_bytes(body)
+        .await
+        .map_err(|err| err.to_string())?;
+    head.uri = Uri::from_str(&format!("http://{}{}", host, head.uri))
+        .expect("if the authority is some then the URL is valid; qed");
+    let req = http_compat::Request::from_parts(head, body);
+    let res = handler.oneshot(req).await.map_err(|err| err.to_string())?;
+
+    let is_json = matches!(
+        res.body(),
+        ResponseBody::GraphQL(_) | ResponseBody::RawJSON(_)
+    );
+
+    let mut res = res.map(|body| match body {
+        ResponseBody::GraphQL(res) => {
+            Bytes::from(serde_json::to_vec(&res).expect("responsebody is serializable; qed"))
+        }
+        ResponseBody::RawJSON(res) => {
+            Bytes::from(serde_json::to_vec(&res).expect("responsebody is serializable; qed"))
+        }
+        ResponseBody::Text(res) => Bytes::from(res),
+    });
+
+    if is_json {
+        res.headers_mut().insert(
+            http::header::CONTENT_TYPE,
+            HeaderValue::from_static("application/json"),
+        );
+    }
+
+    Ok::<_, String>(res)
+}
+
+async fn redirect_or_run_graphql_operation(
+    host: Host,
+    RawQuery(query): RawQuery,
+    headers: HeaderMap,
+    OriginalUri(uri): OriginalUri,
+    Extension(service): Extension<BufferedService>,
+    display_landing_page: bool,
+) -> impl IntoResponse {
+    if headers.get("accept").map(prefers_html).unwrap_or_default() && display_landing_page {
+        return display_home_page().into_response();
+    }
+
+    if query.is_some() {
```

Comment:
```suggestion
    if let Some(request) = query.and_then(|q| graphql::Request::from_urlencoded_query(q).ok()) {
```
(not sure it's really more readable)

Reply:
Looks good !